src/cutadapt/_kmer_finder.pyx

```diff
@@ -223,39 +221,38 @@ cdef void set_masks(bitmask_t *needle_mask, size_t pos, const char *chars):
     cdef char c
     cdef size_t i
     for i in range(strlen(chars)):
-        needle_mask[<uint8_t>chars[i]] &= ~(<bitmask_t>1ULL << pos)
+        needle_mask[<uint8_t>chars[i]] |= <bitmask_t>1ULL << pos
 
 cdef populate_needle_mask(bitmask_t *needle_mask, const char *needle, size_t needle_length,
                           match_lookup):
     cdef size_t i
     cdef char c
     cdef uint8_t j
-    if needle_length > (sizeof(bitmask_t) * 8 - 1):
+    if needle_length > MAX_WORD_SIZE:
         raise ValueError("The pattern is too long!")
-    memset(needle_mask, 0xff, sizeof(bitmask_t) * BITMASK_INDEX_SIZE)
+    memset(needle_mask, 0, sizeof(bitmask_t) * BITMASK_INDEX_SIZE)
     for i in range(needle_length):
         c = needle[i]
         if c == 0:
             continue
         set_masks(needle_mask, i, match_lookup[c])
 
 
-cdef bint shift_or_multiple_is_present(
+cdef bint shift_and_multiple_is_present(
     const char *haystack,
     size_t haystack_length,
     const bitmask_t *needle_mask,
-    bitmask_t zero_mask,
+    bitmask_t init_mask,
     bitmask_t found_mask):
     cdef:
-        bitmask_t R = zero_mask
+        bitmask_t R = init_mask
         size_t i
 
     for i in range(haystack_length):
         # Update the bit array
-        R |= needle_mask[<uint8_t>haystack[i]]
-        R <<= 1
-        R &= zero_mask
-        if (R & found_mask) != found_mask:
+        R &= needle_mask[<uint8_t>haystack[i]]
```

Comment:
We discussed already via e-mail that the order in which you do the update is different from the textbook way of doing it. Now that I see the code again, I understand why the two are equivalent.

Wouldn’t it work and be closer to the textbook way if you initialized `bitmask_t R = 0` and then moved the two last lines of the loop to become the first:
```python
    for i in range(haystack_length):
        # Update the bit array
        R <<= 1
        R |= init_mask
        R &= needle_mask[<uint8_t>haystack[i]]
	    if R & found_mask:
            return True
```
This results in two extra instructions at the beginning of the first loop, but avoids two extra instructions at the end of the last (in case the pattern wasn’t found). I’ll 

BTW, are you aware of how the shift-and algorithm can be interpreted as the bitparallel implementation of a nondeterministic finite automaton (NFA)? Initializing with zero and starting with the shift is more natural in that interpretation.

Reply:
You are right. It looks much nicer that way. It is done.

> BTW, are you aware of how the shift-and algorithm can be interpreted as the bitparallel implementation of a nondeterministic finite automaton (NFA)? Initializing with zero and starting with the shift is more natural in that interpretation.

Yes, a colleague explained finite-state machines to me as a way of searching for multiple words in a sequence. I have seen shift-and described as such an automaton. 
What I like about it is how it translates to simple machine operations and how easy it is to implement. It is also extremely useful. For sequali I moved the algorithm to SSE2 128-bit vectors and that program now looks for 6 12-letter words simultaneously in each sequence at ridiculous speeds. 
Having said that, the statement after the if clause in sequali is much more elaborate as it needs to record the position at which the adapter-stub was found as well as which of the 6 adapter-stubs was found. Moving the shift and init mask to the front will improbe readibility quite a lot there too. Thanks!